change lcm subscriber system to obey context

siyuanfeng-tri opened the pull request:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/robotlocomotion/drake/5712)
<!-- Reviewable:end -->

siyuanfeng-tri commented:
+@jwnimmer-tri for feature review. thanks.
note that i broke two example demo, which needs to be fixed after https://github.com/RobotLocomotion/drake/pull/5672
i also need to merge https://github.com/RobotLocomotion/drake/pull/5709 once that cleans up the lcm_sub_test. 

---

Review status: 0 of 5 files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/5712#-:-Kgq_Zi0RVs2SxSgA5fY:b7pl5rf)*
<!-- Sent from Reviewable.io -->

jwnimmer-tri commented:
Reviewed 2 of 5 files at r1.
Review status: 2 of 5 files reviewed at latest revision, 3 unresolved discussions.

---

*[drake/systems/lcm/lcm_subscriber_system.cc, line 170 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/5712#-KgqeuyCed-rUaZ3ayJv:-KgqeuyCed-rUaZ3ayJw:btcaaey) ([raw file](https://github.com/robotlocomotion/drake/blob/a3c26eaa3f5768d93e327ba6f9e3662cddada34a/drake/systems/lcm/lcm_subscriber_system.cc#L170)):*
> ```cpp
>   DRAKE_DEMAND(translator_ != nullptr);
>   return std::make_unique<AbstractValues>();
> }
> ```

I've only just started thinking about it, but it seems to me like in both cases (translator or serializer), we could store the "next value" in abstract state and use unrestricted updates in both cases as well.  (Even with the same datatypes as now -- the `Value<my_msg>` from a serializer, or the `BasicVector` from a translator.  If that's true, what's the rough thinking on keeping discrete state around as an option beyond fully abstract state?  Do we need to expose it numerically for feedthrough analysis or anything?

---

*[drake/systems/lcm/lcm_subscriber_system.cc, line 170 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/5712#-Kgqmun6eT-g4KNyMGSR:-Kgqmun6eT-g4KNyMGSS:b3aox3p) ([raw file](https://github.com/robotlocomotion/drake/blob/a3c26eaa3f5768d93e327ba6f9e3662cddada34a/drake/systems/lcm/lcm_subscriber_system.cc#L170)):*
> ```cpp
>   DRAKE_DEMAND(translator_ != nullptr);
>   return std::make_unique<AbstractValues>();
> }
> ```

Relatedly, it seems like we could have the message contents our abstract state be represented as `std::vector<uint8_t>` and then run the serializer / translator step during `DoCalcOutput` like we did before this PR.  Then the unrestricted update would just copy the bytes.  If that's true, what's the rough thinking on doing the deserialization during the event, instead when forming an output?

---

*[drake/systems/lcm/lcm_subscriber_system.h, line 192 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/5712#-KgqgNDhn1-0UvYwB3Y5:-KgqgNDhn1-0UvYwB3Y6:b-6zv468) ([raw file](https://github.com/robotlocomotion/drake/blob/a3c26eaa3f5768d93e327ba6f9e3662cddada34a/drake/systems/lcm/lcm_subscriber_system.h#L192)):*
> ```objc
> 
>   // A condition variable that's signaled every time the handler is called.
>   mutable std::condition_variable received_message_condition_variable_;
> ```

It this used for anything yet?  (If not, is there a link to a dev branch where it is used that I can take a look at?)

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/5712)*
<!-- Sent from Reviewable.io -->

siyuanfeng-tri commented:
Review status: 2 of 5 files reviewed at latest revision, 3 unresolved discussions.

---

*[drake/systems/lcm/lcm_subscriber_system.cc, line 170 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/5712#-KgqeuyCed-rUaZ3ayJv:-KgqqUTiZobRscrhJHF6:b-vtheax) ([raw file](https://github.com/robotlocomotion/drake/blob/a3c26eaa3f5768d93e327ba6f9e3662cddada34a/drake/systems/lcm/lcm_subscriber_system.cc#L170)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary><blockquote>

I've only just started thinking about it, but it seems to me like in both cases (translator or serializer), we could store the "next value" in abstract state and use unrestricted updates in both cases as well.  (Even with the same datatypes as now -- the `Value<my_msg>` from a serializer, or the `BasicVector` from a translator.  If that's true, what's the rough thinking on keeping discrete state around as an option beyond fully abstract state?  Do we need to expose it numerically for feedthrough analysis or anything?
</blockquote></details>

It's as stupid as I can wrap a basic vector in AbstractValue because it's not assignable. 

---

*[drake/systems/lcm/lcm_subscriber_system.cc, line 170 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/5712#-Kgqmun6eT-g4KNyMGSR:-Kgqqat4Ej1djZP-aB6f:b5kvyb6) ([raw file](https://github.com/robotlocomotion/drake/blob/a3c26eaa3f5768d93e327ba6f9e3662cddada34a/drake/systems/lcm/lcm_subscriber_system.cc#L170)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary><blockquote>

Relatedly, it seems like we could have the message contents our abstract state be represented as `std::vector<uint8_t>` and then run the serializer / translator step during `DoCalcOutput` like we did before this PR.  Then the unrestricted update would just copy the bytes.  If that's true, what's the rough thinking on doing the deserialization during the event, instead when forming an output?
</blockquote></details>

i think about this a bit. i am leaning slightly towards the current setup because of speed, which is likely to change with the new cache. 
you only really do state update once, thus 1 call to decode, but you could be eval output many many time, thus decoding many time.s  

---

*[drake/systems/lcm/lcm_subscriber_system.h, line 192 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/5712#-KgqgNDhn1-0UvYwB3Y5:-KgqqtkyGBFcFSeIJXNc:bfeo2wn) ([raw file](https://github.com/robotlocomotion/drake/blob/a3c26eaa3f5768d93e327ba6f9e3662cddada34a/drake/systems/lcm/lcm_subscriber_system.h#L192)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary><blockquote>

It this used for anything yet?  (If not, is there a link to a dev branch where it is used that I can take a look at?)
</blockquote></details>

i nuked it from this pr in r2. 

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/5712)*
<!-- Sent from Reviewable.io -->

jwnimmer-tri commented:
Review status: 2 of 5 files reviewed at latest revision, 2 unresolved discussions.

---

*[drake/systems/lcm/lcm_subscriber_system.cc, line 170 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/5712#-KgqeuyCed-rUaZ3ayJv:-KgqsAXVh4LADFzKJG4a:bprwqzz) ([raw file](https://github.com/robotlocomotion/drake/blob/a3c26eaa3f5768d93e327ba6f9e3662cddada34a/drake/systems/lcm/lcm_subscriber_system.cc#L170)):*
<details><summary><i>Previously, siyuanfeng-tri wrote…</i></summary><blockquote>

It's as stupid as I can wrap a basic vector in AbstractValue because it's not assignable. 
</blockquote></details>

`VectorValue<double>` is-a `AbstractValue` and could be put into abstract state. Granted that would be somewhat unconventional, but possibly worth it for code de-duplication?

---

*[drake/systems/lcm/lcm_subscriber_system.cc, line 170 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/5712#-Kgqmun6eT-g4KNyMGSR:-KgqszlkwimPHBx6jLl_:ba3eey1) ([raw file](https://github.com/robotlocomotion/drake/blob/a3c26eaa3f5768d93e327ba6f9e3662cddada34a/drake/systems/lcm/lcm_subscriber_system.cc#L170)):*
<details><summary><i>Previously, siyuanfeng-tri wrote…</i></summary><blockquote>

i think about this a bit. i am leaning slightly towards the current setup because of speed, which is likely to change with the new cache. 
you only really do state update once, thus 1 call to decode, but you could be eval output many many time, thus decoding many time.s  
</blockquote></details>

The version on master already deserializes on every `DoCalcOutput`; has that been a performance problem?  If yes then I'm fine with documenting this choice and moving the deserialization into the event handler like we do here, but we just need to be sure to capture in comments _why_ that choice was made.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/5712)*
<!-- Sent from Reviewable.io -->

siyuanfeng-tri commented:
Review status: 2 of 5 files reviewed at latest revision, 2 unresolved discussions.

---

*[drake/systems/lcm/lcm_subscriber_system.cc, line 170 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/5712#-Kgqmun6eT-g4KNyMGSR:-KgqvCVQIjuKdt3M0mD6:bawn8ue) ([raw file](https://github.com/robotlocomotion/drake/blob/a3c26eaa3f5768d93e327ba6f9e3662cddada34a/drake/systems/lcm/lcm_subscriber_system.cc#L170)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary><blockquote>

The version on master already deserializes on every `DoCalcOutput`; has that been a performance problem?  If yes then I'm fine with documenting this choice and moving the deserialization into the event handler like we do here, but we just need to be sure to capture in comments _why_ that choice was made.
</blockquote></details>

i don't think there is a performance issue at the moment for anything we have being doing. but i just don't think that was a good choice to begin with. so i made it this way. i can imagine some gigantic point cloud message is being decoded here. now you could argue that if decode takes a while, we shouldn't do decoding with the lock, which i agree. 

---

*[drake/systems/lcm/lcm_subscriber_system.cc, line 170 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/5712#-KgqeuyCed-rUaZ3ayJv:-Kgqw2hG5FI70gLjZ_UY:b-3284d6) ([raw file](https://github.com/robotlocomotion/drake/blob/a3c26eaa3f5768d93e327ba6f9e3662cddada34a/drake/systems/lcm/lcm_subscriber_system.cc#L170)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary><blockquote>

`VectorValue<double>` is-a `AbstractValue` and could be put into abstract state. Granted that would be somewhat unconventional, but possibly worth it for code de-duplication?
</blockquote></details>

i also tried it a bit. but didn't get it to work correctly. granted i didn't try very hard, and i think i failed to get the CustomVector thing working in the test case. it had nothing to do with vector value per say, but it was a pain in the ass to work with because of the extra layer of wrapping. the code was pretty much unreadable. 
even if i put everything in abstract state, it still doesn't really solve the condition on translator_ vs serializer_, which is the fundamental cause of the basicvector xor abstract value implementation. how much of the code base is relying on the basic vector version? do you have a feeling? 

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/5712)*
<!-- Sent from Reviewable.io -->

siyuanfeng-tri commented:
cc @sammy-tri becaue i broke iiwa_controller.cc here, but i promisie i will fix it before merging this.

---







*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/5712#-:-KgrHoZLtDCiGnThOUbi:bnui8rp)*
<!-- Sent from Reviewable.io -->

jwnimmer-tri commented:
Review status: 2 of 5 files reviewed at latest revision, 2 unresolved discussions, some commit checks failed.

---

*[drake/systems/lcm/lcm_subscriber_system.cc, line 170 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/5712#-KgqeuyCed-rUaZ3ayJv:-KgturAhKWZ7LwSOMjJm:b-r3npos) ([raw file](https://github.com/robotlocomotion/drake/blob/a3c26eaa3f5768d93e327ba6f9e3662cddada34a/drake/systems/lcm/lcm_subscriber_system.cc#L170)):*
<details><summary><i>Previously, siyuanfeng-tri wrote…</i></summary><blockquote>

i also tried it a bit. but didn't get it to work correctly. granted i didn't try very hard, and i think i failed to get the CustomVector thing working in the test case. it had nothing to do with vector value per say, but it was a pain in the ass to work with because of the extra layer of wrapping. the code was pretty much unreadable. 
even if i put everything in abstract state, it still doesn't really solve the condition on translator_ vs serializer_, which is the fundamental cause of the basicvector xor abstract value implementation. how much of the code base is relying on the basic vector version? do you have a feeling? 
</blockquote></details>

IIRC Pat has also suggested making LcmSubscriberSystem output _only_ the abstract `Value<my_message>` and not have an option for `BasicVector`.  I am fine if you want to change it to work that way (separate PR before or after this one, maybe?).  _But_ we do still need the `BasicVector` concept, so we'll need to simultaneously offer a direct-feedthrough `System` that does `Value<my_message>` to `MyCustomVector` using a translator, and port users to insert that (another PR).

In any case, the question atop this thread was "Please illuminate why we are using two different kinds of state", which you've explained in this thread.  If you want to keep the design this way, it's fine by me and I'll do the review to make sure it's correct and has enough unit tests to cover all of the branches.

---

*[drake/systems/lcm/lcm_subscriber_system.cc, line 170 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/5712#-Kgqmun6eT-g4KNyMGSR:-Kgtw--IzyEp6kgoHZKl:b955cyk) ([raw file](https://github.com/robotlocomotion/drake/blob/a3c26eaa3f5768d93e327ba6f9e3662cddada34a/drake/systems/lcm/lcm_subscriber_system.cc#L170)):*
<details><summary><i>Previously, siyuanfeng-tri wrote…</i></summary><blockquote>

i don't think there is a performance issue at the moment for anything we have being doing. but i just don't think that was a good choice to begin with. so i made it this way. i can imagine some gigantic point cloud message is being decoded here. now you could argue that if decode takes a while, we shouldn't do decoding with the lock, which i agree. 
</blockquote></details>

Okay, so that's an acceptable explanation, but needs to turn into code comments if you want to keep it this way.  It's not obvious to a naive reader why you've chosen the extra complexity for the possible performance benefit here.

FWIW I don't think the performance argument actually matters, because DoCalcOutput shouldn't be recomputing its results, ever -- either the Simulator should not be calling it over and over, or it should use the cache to answer the queries (probably in EvalOutput, which we wouldn't even see).  So personally I wouldn't make this code more obscure by manually caching the transformation, but I'll allow it if you want to keep this way.  Again, I'll make sure the code and tests are thorough with respect to this complexity when I begin the full review.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/5712)*
<!-- Sent from Reviewable.io -->

sammy-tri commented:
I can live with the iiwa_controller breakage.  I only rarely bother to run it, and I've been on the verge of deleting it anyway.

---

Review status: 2 of 5 files reviewed at latest revision, 2 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/5712#-:-KguWsUO2tM-BrQh4CTC:b-21vran)*
<!-- Sent from Reviewable.io -->

jwnimmer-tri commented:
So for process clarity: Siyuan, for the moment I'm not reading more of this PR.  If you want to keep the design / model more or less like it is now, re-ping me and I'll resume reading the rest of the code and test (and I'll eventually expect a few comments to eventually show up per the two open discussions, but I can still review the rest now anyway).  If I don't hear anything, I'll assume a new push is bring worked on and then I'll resume review after that happens.

---







*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/5712#-:-KgvAwXwfocmqDd-zzNQ:b-8vp8sc)*
<!-- Sent from Reviewable.io -->

siyuanfeng-tri commented:
hey Jeremy, i am not planning to do any major changes. i will probably put the conversation in some form of doc later tonight. so you can proceed, meanwhile there is the drake viz pr i just assigned to you :| thanks!

---







*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/5712#-:-KgvaPMvFV4cjqITYeOq:b-shahtj)*
<!-- Sent from Reviewable.io -->

jwnimmer-tri commented:
Checkpoint.

---

Reviewed 2 of 3 files at r2.
Review status: 4 of 5 files reviewed at latest revision, 13 unresolved discussions.

---

*[drake/systems/lcm/lcm_subscriber_system.cc, line 93 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/5712#-KgvqgMhaLgWm5fzKab_:-KgvqgMhaLgWm5fzKaba:b-eim53w) ([raw file](https://github.com/robotlocomotion/drake/blob/d7200024fd5166c23d078743d0a249d2951e7fc2/drake/systems/lcm/lcm_subscriber_system.cc#L93)):*
> ```cpp
>         discrete_state->get_mutable_discrete_state(state_index_msg_));
>     discrete_state->get_mutable_discrete_state(state_index_msg_ctr_)
>         ->SetAtIndex(0, received_message_count_);
> ```

Please comment why, in the case when there is no message received yet, we should leave the `state_index_msg_ctr_` state unset?  It naively seems to me like this statement should happen outside of the `if` guard.

Ditto for the similar code in `Process...` for `msg_ctr` below.

---

*[drake/systems/lcm/lcm_subscriber_system.cc, line 130 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/5712#-KgvsXa4N3r28nSNLMiK:-KgvsXa5A8cS5vcqfuDx:b91xhwt) ([raw file](https://github.com/robotlocomotion/drake/blob/d7200024fd5166c23d078743d0a249d2951e7fc2/drake/systems/lcm/lcm_subscriber_system.cc#L130)):*
> ```cpp
>   // Has a new message. Schedule an update event.
>   if (last_message_count != received_message_count_) {
>     events->time = context.get_time() + 0.0001;
> ```

The 1e-4 is a fudge to placate the Simulator, yes? (Because the Simulator complains if return return "now", I think?) If so, please add a TODO here that we should fix the Simulator and remove the fudge.

(I guess given the several times this TODO will appear now, perhaps a simple issue number cited here would be more efficient documentation.)

---

*[drake/systems/lcm/lcm_subscriber_system.cc, line 134 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/5712#-KgvsxxzPbTue9OpssF_:-KgvsxxzPbTue9OpssFa:bbf2vyi) ([raw file](https://github.com/robotlocomotion/drake/blob/d7200024fd5166c23d078743d0a249d2951e7fc2/drake/systems/lcm/lcm_subscriber_system.cc#L134)):*
> ```cpp
>     if (translator_ == nullptr)
>       event.action =
>           DiscreteEvent<double>::ActionType::kUnrestrictedUpdateAction;
> ```

BTW Multi-line if-else bodies require curly braces (on both bodies).

---

*[drake/systems/lcm/lcm_subscriber_system.cc, line 153 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/5712#-KgvpRQE4vWjoaGRoqHD:-KgvpRQE4vWjoaGRoqHE:b-f602wn) ([raw file](https://github.com/robotlocomotion/drake/blob/d7200024fd5166c23d078743d0a249d2951e7fc2/drake/systems/lcm/lcm_subscriber_system.cc#L153)):*
> ```cpp
>     discrete_state_vec[state_index_msg_ctr_] =
>         std::make_unique<BasicVector<double>>(1);
>     discrete_state_vec[state_index_msg_ctr_]->SetAtIndex(0, 0);
> ```

I'm not clear on why setting to zero is required here?  Typically `Allocate...` overrides only need to worry about types and dimensions, not content.

---

*[drake/systems/lcm/lcm_subscriber_system.cc, line 158 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/5712#-Kgvp9_5kkFritdd76NG:-Kgvp9_6jshiN1rNTGy1:bpjcklv) ([raw file](https://github.com/robotlocomotion/drake/blob/d7200024fd5166c23d078743d0a249d2951e7fc2/drake/systems/lcm/lcm_subscriber_system.cc#L158)):*
> ```cpp
>   }
>   DRAKE_DEMAND(serializer_ != nullptr);
>   return std::make_unique<DiscreteState<double>>();
> ```

FYI In cases where we didn't need an override, sometimes it's more robust to call the base class, versus reimplement it; so `return LeafSystem<double>::AllocateDiscreteState();` in this case.  Ditto for `AllocateAbstractState`'s second return statement below, and possibly even the `DoCalcNextUpdate` case of nothing to do.

---

*[drake/systems/lcm/lcm_subscriber_system.h, line 26 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/5712#-KgvkBhGkoxjIEHubsOf:-KgvkBhGkoxjIEHubsOg:b-bel2oz) ([raw file](https://github.com/robotlocomotion/drake/blob/d7200024fd5166c23d078743d0a249d2951e7fc2/drake/systems/lcm/lcm_subscriber_system.h#L26)):*
> ```objc
>  * System<double>'s port. This class stores the most recently processed Lcm
>  * message in its Context. CalcOutput() merely copies the message from the
>  * Context to the output. To process a Lcm message, the user need to call
> ```

BTW two nits:
s/user need/user needs/;
s/Lcm/LCM/    (also in the subsequent line)

---

*[drake/systems/lcm/lcm_subscriber_system.h, line 30 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/5712#-Kgvkr6hr0Q0B_da22Ki:-Kgvkr6hr0Q0B_da22Kj:b-dz1kg3) ([raw file](https://github.com/robotlocomotion/drake/blob/d7200024fd5166c23d078743d0a249d2951e7fc2/drake/systems/lcm/lcm_subscriber_system.h#L30)):*
> ```objc
>  * has arrived. The message is then processed and stored in the Context by
>  * calling CalcDiscreteVariableUpdates() or CalcUnrestrictedUpdate() depending
>  * on the output type.
> ```

So I think walking through the mechanics of how this works is great, but perhaps it should be more like an implementation-styled comment than an API overview comment?  The details of which methods do what doesn't seem important to the user.

I think in the overview, for the user perhaps the most important concepts are that:
(1) the System subscribes to an LCM channel upon construction;
(2) when a new message appears asynchronously over LCM, the System declares an unrestricted update event at the soonest possible simulation time;
(3) the event updates the System's State (within the Context) to be the most recently received message;
(4) the System's output is always the same as its State;

... so the bottom line is that the System's output is always a pure function of its State, the State only changes during an update event as if the simulation was paused.

Or something like that.

---

*[drake/systems/lcm/lcm_subscriber_system.h, line 166 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/5712#-KgvnTyNwST92aJ5xWD2:-KgvnTyNwST92aJ5xWD3:bq0vwp4) ([raw file](https://github.com/robotlocomotion/drake/blob/d7200024fd5166c23d078743d0a249d2951e7fc2/drake/systems/lcm/lcm_subscriber_system.h#L166)):*
> ```objc
> 
>   // Callback entry point from LCM into this class. Also wakes up one thread
>   // block on notification_ if it's not nullptr.
> ```

BTW The second sentence is a leftover, I think?

---

*[drake/systems/lcm/lcm_subscriber_system.h, line 181 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/5712#-KgvnYs48A5B_PDhP5Pw:-KgvnYs48A5B_PDhP5Px:b-6mf5pk) ([raw file](https://github.com/robotlocomotion/drake/blob/d7200024fd5166c23d078743d0a249d2951e7fc2/drake/systems/lcm/lcm_subscriber_system.h#L181)):*
> ```objc
>   const std::unique_ptr<SerializerInterface> serializer_;
> 
>   // The mutex that guards received_message_.
> ```

It also guards `received_message_count_`, I think.

---

*[drake/systems/lcm/lcm_subscriber_system.h, line 192 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/5712#-KgvnrIbyimHCBzvmS1w:-KgvnrIbyimHCBzvmS1x:b-5gkb6x) ([raw file](https://github.com/robotlocomotion/drake/blob/d7200024fd5166c23d078743d0a249d2951e7fc2/drake/systems/lcm/lcm_subscriber_system.h#L192)):*
> ```objc
>   // Index 0 is the message, 1 is the message counter.
>   const int state_index_msg_{0};
>   const int state_index_msg_ctr_{1};
> ```

BTW This two seem constant for all time; should they be namespace constants in the `.cc` file, instead of member fields?

---

*[drake/systems/lcm/lcm_subscriber_system.h, line 192 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/5712#-Kgvq0NNOe9fLkVT2nA_:-Kgvq0NNOe9fLkVT2nAa:bbgerke) ([raw file](https://github.com/robotlocomotion/drake/blob/d7200024fd5166c23d078743d0a249d2951e7fc2/drake/systems/lcm/lcm_subscriber_system.h#L192)):*
> ```objc
>   // Index 0 is the message, 1 is the message counter.
>   const int state_index_msg_{0};
>   const int state_index_msg_ctr_{1};
> ```

I think `ctr` is an abbreviation formed by deleting letters from a word?  If so, that is forbidden by cppguide.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/5712#-:-Kgvu1DrjI8pLLNZ0wj0:by96gdd)*
<!-- Sent from Reviewable.io -->

siyuanfeng-tri commented:
Review status: 2 of 5 files reviewed at latest revision, 13 unresolved discussions.

---

*[drake/systems/lcm/lcm_subscriber_system.cc, line 170 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/5712#-KgqeuyCed-rUaZ3ayJv:-Kgu3qoktGB1-7sIN4WL:b8hcset) ([raw file](https://github.com/robotlocomotion/drake/blob/a3c26eaa3f5768d93e327ba6f9e3662cddada34a/drake/systems/lcm/lcm_subscriber_system.cc#L170)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary><blockquote>

IIRC Pat has also suggested making LcmSubscriberSystem output _only_ the abstract `Value<my_message>` and not have an option for `BasicVector`.  I am fine if you want to change it to work that way (separate PR before or after this one, maybe?).  _But_ we do still need the `BasicVector` concept, so we'll need to simultaneously offer a direct-feedthrough `System` that does `Value<my_message>` to `MyCustomVector` using a translator, and port users to insert that (another PR).

In any case, the question atop this thread was "Please illuminate why we are using two different kinds of state", which you've explained in this thread.  If you want to keep the design this way, it's fine by me and I'll do the review to make sure it's correct and has enough unit tests to cover all of the branches.
</blockquote></details>

ok. in that case i am leaning towards pure abstract value output, and a translator block afterwords. i will leave it for another day. 

---

*[drake/systems/lcm/lcm_subscriber_system.cc, line 170 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/5712#-Kgqmun6eT-g4KNyMGSR:-Kgu2qiQNnl3FpxjYNxz:b-dpuoel) ([raw file](https://github.com/robotlocomotion/drake/blob/a3c26eaa3f5768d93e327ba6f9e3662cddada34a/drake/systems/lcm/lcm_subscriber_system.cc#L170)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary><blockquote>

Okay, so that's an acceptable explanation, but needs to turn into code comments if you want to keep it this way.  It's not obvious to a naive reader why you've chosen the extra complexity for the possible performance benefit here.

FWIW I don't think the performance argument actually matters, because DoCalcOutput shouldn't be recomputing its results, ever -- either the Simulator should not be calling it over and over, or it should use the cache to answer the queries (probably in EvalOutput, which we wouldn't even see).  So personally I wouldn't make this code more obscure by manually caching the transformation, but I'll allow it if you want to keep this way.  Again, I'll make sure the code and tests are thorough with respect to this complexity when I begin the full review.
</blockquote></details>

i agree that should be properly cached eventually, and the computation should happen once. but before i am seeing that, i am going to avoid recomputation as much as possible. added a block comment.

---

*[drake/systems/lcm/lcm_subscriber_system.cc, line 93 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/5712#-KgvqgMhaLgWm5fzKab_:-KgyhKIARJllZ9cDk7Tj:b-m9jgpc) ([raw file](https://github.com/robotlocomotion/drake/blob/d7200024fd5166c23d078743d0a249d2951e7fc2/drake/systems/lcm/lcm_subscriber_system.cc#L93)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary><blockquote>

Please comment why, in the case when there is no message received yet, we should leave the `state_index_msg_ctr_` state unset?  It naively seems to me like this statement should happen outside of the `if` guard.

Ditto for the similar code in `Process...` for `msg_ctr` below.
</blockquote></details>

opps. this is code copy pasted from a couple days ago when i wasn't aware of the set default state func. 
i used to set it in the allocation. i changed it to conform with the allocate / set pair. 

---

*[drake/systems/lcm/lcm_subscriber_system.cc, line 130 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/5712#-KgvsXa4N3r28nSNLMiK:-Kgydhmmhu3HjeQnH4rA:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/d7200024fd5166c23d078743d0a249d2951e7fc2/drake/systems/lcm/lcm_subscriber_system.cc#L130)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary><blockquote>

The 1e-4 is a fudge to placate the Simulator, yes? (Because the Simulator complains if return return "now", I think?) If so, please add a TODO here that we should fix the Simulator and remove the fudge.

(I guess given the several times this TODO will appear now, perhaps a simple issue number cited here would be more efficient documentation.)
</blockquote></details>

Done.

---

*[drake/systems/lcm/lcm_subscriber_system.cc, line 134 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/5712#-KgvsxxzPbTue9OpssF_:-KgydnTQEjC6gdi7z8cy:b3oh8n) ([raw file](https://github.com/robotlocomotion/drake/blob/d7200024fd5166c23d078743d0a249d2951e7fc2/drake/systems/lcm/lcm_subscriber_system.cc#L134)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary><blockquote>

BTW Multi-line if-else bodies require curly braces (on both bodies).
</blockquote></details>


Done.

---

*[drake/systems/lcm/lcm_subscriber_system.cc, line 153 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/5712#-KgvpRQE4vWjoaGRoqHD:-KgyfPOGdGuF5cQyufKO:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/d7200024fd5166c23d078743d0a249d2951e7fc2/drake/systems/lcm/lcm_subscriber_system.cc#L153)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary><blockquote>

I'm not clear on why setting to zero is required here?  Typically `Allocate...` overrides only need to worry about types and dimensions, not content.
</blockquote></details>

Done.

---

*[drake/systems/lcm/lcm_subscriber_system.cc, line 158 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/5712#-Kgvp9_5kkFritdd76NG:-Kgyfki_kcXulVBchJZP:b-etnlwc) ([raw file](https://github.com/robotlocomotion/drake/blob/d7200024fd5166c23d078743d0a249d2951e7fc2/drake/systems/lcm/lcm_subscriber_system.cc#L158)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary><blockquote>

FYI In cases where we didn't need an override, sometimes it's more robust to call the base class, versus reimplement it; so `return LeafSystem<double>::AllocateDiscreteState();` in this case.  Ditto for `AllocateAbstractState`'s second return statement below, and possibly even the `DoCalcNextUpdate` case of nothing to do.
</blockquote></details>

good point. 

---

*[drake/systems/lcm/lcm_subscriber_system.h, line 26 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/5712#-KgvkBhGkoxjIEHubsOf:-Kgyaz7b_pbtzCdx5NEs:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/d7200024fd5166c23d078743d0a249d2951e7fc2/drake/systems/lcm/lcm_subscriber_system.h#L26)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary><blockquote>

BTW two nits:
s/user need/user needs/;
s/Lcm/LCM/    (also in the subsequent line)
</blockquote></details>

Done.

---

*[drake/systems/lcm/lcm_subscriber_system.h, line 30 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/5712#-Kgvkr6hr0Q0B_da22Ki:-KgydH7yrupHdwtkJcj8:bia13pb) ([raw file](https://github.com/robotlocomotion/drake/blob/d7200024fd5166c23d078743d0a249d2951e7fc2/drake/systems/lcm/lcm_subscriber_system.h#L30)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary><blockquote>

So I think walking through the mechanics of how this works is great, but perhaps it should be more like an implementation-styled comment than an API overview comment?  The details of which methods do what doesn't seem important to the user.

I think in the overview, for the user perhaps the most important concepts are that:
(1) the System subscribes to an LCM channel upon construction;
(2) when a new message appears asynchronously over LCM, the System declares an unrestricted update event at the soonest possible simulation time;
(3) the event updates the System's State (within the Context) to be the most recently received message;
(4) the System's output is always the same as its State;

... so the bottom line is that the System's output is always a pure function of its State, the State only changes during an update event as if the simulation was paused.

Or something like that.
</blockquote></details>

updated the comments. 

---

*[drake/systems/lcm/lcm_subscriber_system.h, line 166 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/5712#-KgvnTyNwST92aJ5xWD2:-KgyahENyJPO-QDL9YHj:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/d7200024fd5166c23d078743d0a249d2951e7fc2/drake/systems/lcm/lcm_subscriber_system.h#L166)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary><blockquote>

BTW The second sentence is a leftover, I think?
</blockquote></details>

Done.

---

*[drake/systems/lcm/lcm_subscriber_system.h, line 181 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/5712#-KgvnYs48A5B_PDhP5Pw:-KgyacpCjZzO7KPPbC31:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/d7200024fd5166c23d078743d0a249d2951e7fc2/drake/systems/lcm/lcm_subscriber_system.h#L181)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary><blockquote>

It also guards `received_message_count_`, I think.
</blockquote></details>

Done.

---

*[drake/systems/lcm/lcm_subscriber_system.h, line 192 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/5712#-KgvnrIbyimHCBzvmS1w:-KgyaXozfPlzmmbMo0cI:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/d7200024fd5166c23d078743d0a249d2951e7fc2/drake/systems/lcm/lcm_subscriber_system.h#L192)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary><blockquote>

BTW This two seem constant for all time; should they be namespace constants in the `.cc` file, instead of member fields?
</blockquote></details>

Done.

---

*[drake/systems/lcm/lcm_subscriber_system.h, line 192 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/5712#-Kgvq0NNOe9fLkVT2nA_:-KgyaY7zpSeTYDAdzON9:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/d7200024fd5166c23d078743d0a249d2951e7fc2/drake/systems/lcm/lcm_subscriber_system.h#L192)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary><blockquote>

I think `ctr` is an abbreviation formed by deleting letters from a word?  If so, that is forbidden by cppguide.
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/5712)*
<!-- Sent from Reviewable.io -->

jwnimmer-tri commented:
Checkpoint.  In good shape, though I haven't started to look at the unit testing yet.  Might be an okay time to bring in a platform reviewer.

---

Reviewed 2 of 2 files at r3.
Review status: 4 of 5 files reviewed at latest revision, 2 unresolved discussions, some commit checks failed.

---

*[drake/systems/lcm/lcm_subscriber_system.cc, line 20 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/5712#-KgyuEND-qE9FAaQzrm_:-KgyuENEvIuc0UFM0dGv:bsimao0) ([raw file](https://github.com/robotlocomotion/drake/blob/d7a3a94d41d9afa9dd6580485eb91af3c4d6f6f8/drake/systems/lcm/lcm_subscriber_system.cc#L20)):*
> ```cpp
> namespace {
>   constexpr int kStateIndexMessage = 0;
>   constexpr int kStateIndexMessageCount = 1;
> ```

BTW Should not be indented.

---

*[drake/systems/lcm/lcm_subscriber_system.h, line 34 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/5712#-KgytwWTw4X_uYkiZUon:-KgytwWUvEtSQIDHXDbb:b-sgufok) ([raw file](https://github.com/robotlocomotion/drake/blob/d7a3a94d41d9afa9dd6580485eb91af3c4d6f6f8/drake/systems/lcm/lcm_subscriber_system.h#L34)):*
> ```objc
>  * has arrived. The message is then processed and stored in the Context by
>  * calling CalcDiscreteVariableUpdates() or CalcUnrestrictedUpdate() depending
>  * on the output type.
> ```

I guess I still feel like this will be misleading for people reading the API overview.  The user doesn't need to do any of this; the Simulator does it for them.  Maybe somehow clarify that you're explaining what the Simulator does, and this text only applies if you're not using it for some reason?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/5712#-:-Kgyvpk7MKatQVNuMFB4:bt4ht2v)*
<!-- Sent from Reviewable.io -->

jwnimmer-tri commented:
Reviewed 1 of 1 files at r4.
Review status: 4 of 5 files reviewed at latest revision, 1 unresolved discussion.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/5712)*
<!-- Sent from Reviewable.io -->

siyuanfeng-tri commented:
Review status: 4 of 5 files reviewed at latest revision, 2 unresolved discussions.

---

*[drake/systems/lcm/lcm_subscriber_system.cc, line 20 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/5712#-KgyuEND-qE9FAaQzrm_:-KgyyLTSbTTlbc0liOy_:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/d7a3a94d41d9afa9dd6580485eb91af3c4d6f6f8/drake/systems/lcm/lcm_subscriber_system.cc#L20)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary><blockquote>

BTW Should not be indented.
</blockquote></details>

Done.

---

*[drake/systems/lcm/lcm_subscriber_system.h, line 34 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/5712#-KgytwWTw4X_uYkiZUon:-KgyySRW9btlqOZjNSf3:bdsepib) ([raw file](https://github.com/robotlocomotion/drake/blob/d7a3a94d41d9afa9dd6580485eb91af3c4d6f6f8/drake/systems/lcm/lcm_subscriber_system.h#L34)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary><blockquote>

I guess I still feel like this will be misleading for people reading the API overview.  The user doesn't need to do any of this; the Simulator does it for them.  Maybe somehow clarify that you're explaining what the Simulator does, and this text only applies if you're not using it for some reason?
</blockquote></details>

good point. added more comment.s 

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/5712)*
<!-- Sent from Reviewable.io -->

jwnimmer-tri commented:
Reviewed 1 of 3 files at r2, 1 of 1 files at r5.
Review status: all files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/5712)*
<!-- Sent from Reviewable.io -->

siyuanfeng-tri commented:
@jwnimmer-tri should i find another feature or platform reviewer? 

---







*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/5712#-:-Kgz3M6MaYc7BaibC00H:bq2t9rm)*
<!-- Sent from Reviewable.io -->

jwnimmer-tri commented:
Yes, this needs a second reviewer.
+@ggould-tri for platform review and design re-review, please.

---

Review status: all files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/5712#-:-Kgz3fj9qvQ1SZjpiiI8:bjhkaid)*
<!-- Sent from Reviewable.io -->

jwnimmer-tri commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> I re-read the changes and tests, and think it's good.

---

Review status: all files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/5712#-:-Kgz9BAPe4tN3_-MrAOh:b-p2t0s7)*
<!-- Sent from Reviewable.io -->

siyuanfeng-tri commented:
thanks, much appreciated. 

---







*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/5712#-:-Kgz9RW66Q51Q236pyEZ:bmbd33h)*
<!-- Sent from Reviewable.io -->

ggould-tri commented:
Reviewed 2 of 5 files at r1, 1 of 3 files at r2, 1 of 1 files at r4, 1 of 1 files at r5.
Review status: all files reviewed at latest revision, 5 unresolved discussions.

---

*[drake/systems/lcm/lcm_subscriber_system.cc, line 31 at r5](https://reviewable.io:443/reviews/robotlocomotion/drake/5712#-KgzFaPS_B_V-2H0Hv8g:-KgzFaPS_B_V-2H0Hv8h:b2eca31) ([raw file](https://github.com/robotlocomotion/drake/blob/891cd64b45ff65b94a1cead2da4a81ad56b53065/drake/systems/lcm/lcm_subscriber_system.cc#L31)):*
> ```cpp
> // the old code yet, but still can supply the AbstractValue port for new code.
> 
> // TODO(siyuan): Probably should only AbstractValue output, and change / move
> ```

FYI first phrase of the sentence is missing a verb.

---

*[drake/systems/lcm/lcm_subscriber_system.cc, line 86 at r5](https://reviewable.io:443/reviews/robotlocomotion/drake/5712#-KgzFlwaUE_i0dU5tFVz:-KgzFlwaUE_i0dU5tFW-:b-dsanku) ([raw file](https://github.com/robotlocomotion/drake/blob/891cd64b45ff65b94a1cead2da4a81ad56b53065/drake/systems/lcm/lcm_subscriber_system.cc#L86)):*
> ```cpp
> // state is to avoid repeated decoding in DoCalcOutput for the latter case.
> // However, this computational concern will not exist once caching is properly
> // implemented and in place. Same for ProcessMessageAndStoreToAbstractState()
> ```

BTW, is there an issue number that this can reference so that it will be clear when it can be resolved?

---

*[drake/systems/lcm/lcm_subscriber_system.cc, line 134 at r5](https://reviewable.io:443/reviews/robotlocomotion/drake/5712#-KgzGX6lYsxu0QSx7axu:-KgzGX6lYsxu0QSx7axv:b9c6lvj) ([raw file](https://github.com/robotlocomotion/drake/blob/891cd64b45ff65b94a1cead2da4a81ad56b53065/drake/systems/lcm/lcm_subscriber_system.cc#L134)):*
> ```cpp
>   // Has a new message. Schedule an update event.
>   if (last_message_count != received_message_count_) {
>     // TODO(siyuan): should be now once #5725 is resolved.
> ```

FYI This will be clearer to any future, post-5725 reader if you make it clear that "now" here is a name, ie change it to now() or whatever the actual call should be (because 5725 may be something that takes a long time to resolve and you may not be the person who fixes this code then).

---

*[drake/systems/lcm/lcm_subscriber_system.h, line 150 at r5](https://reviewable.io:443/reviews/robotlocomotion/drake/5712#-KgzIDEzi4bu2Zm48mCm:-KgzIDF-yyk7IxSAQiLD:b7bpe7o) ([raw file](https://github.com/robotlocomotion/drake/blob/891cd64b45ff65b94a1cead2da4a81ad56b53065/drake/systems/lcm/lcm_subscriber_system.h#L150)):*
> ```objc
>       const Context<double>& context,
>       DiscreteState<double>* discrete_state) const override {
>     ProcessMessageAndStoreToDiscreteState(discrete_state);
> ```

BTW I am surprised you didn't get an unused variable warning on `context` here.

---

*[drake/systems/lcm/test/lcm_subscriber_system_test.cc, line 28 at r5](https://reviewable.io:443/reviews/robotlocomotion/drake/5712#-KgzJ7VE7JcMd4snlZ2T:-KgzJ7VE7JcMd4snlZ2U:btqk15h) ([raw file](https://github.com/robotlocomotion/drake/blob/891cd64b45ff65b94a1cead2da4a81ad56b53065/drake/systems/lcm/test/lcm_subscriber_system_test.cc#L28)):*
> ```cpp
>   sub.CalcNextUpdateTime(*context, &actions);
> 
>   if (!actions.events.empty()) {
> ```

It is not clear to me why this is `if` rather than `while`.  Shouldn't all of `actions` be processed?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/5712)*
<!-- Sent from Reviewable.io -->

siyuanfeng-tri commented:
Review status: 3 of 5 files reviewed at latest revision, 4 unresolved discussions.

---

*[drake/systems/lcm/lcm_subscriber_system.cc, line 31 at r5](https://reviewable.io:443/reviews/robotlocomotion/drake/5712#-KgzFaPS_B_V-2H0Hv8g:-KgzNtoC21TQc8bo1LWJ:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/891cd64b45ff65b94a1cead2da4a81ad56b53065/drake/systems/lcm/lcm_subscriber_system.cc#L31)):*
<details><summary><i>Previously, ggould-tri wrote…</i></summary><blockquote>

FYI first phrase of the sentence is missing a verb.
</blockquote></details>

Done.

---

*[drake/systems/lcm/lcm_subscriber_system.cc, line 86 at r5](https://reviewable.io:443/reviews/robotlocomotion/drake/5712#-KgzFlwaUE_i0dU5tFVz:-KgzNyu0CQAqYhx35gPt:b-v4rv9u) ([raw file](https://github.com/robotlocomotion/drake/blob/891cd64b45ff65b94a1cead2da4a81ad56b53065/drake/systems/lcm/lcm_subscriber_system.cc#L86)):*
<details><summary><i>Previously, ggould-tri wrote…</i></summary><blockquote>

BTW, is there an issue number that this can reference so that it will be clear when it can be resolved?
</blockquote></details>

i am not sure.. it's such a well known issue though. 

---

*[drake/systems/lcm/lcm_subscriber_system.cc, line 134 at r5](https://reviewable.io:443/reviews/robotlocomotion/drake/5712#-KgzGX6lYsxu0QSx7axu:-KgzOEfTbB_qYHfm2MFm:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/891cd64b45ff65b94a1cead2da4a81ad56b53065/drake/systems/lcm/lcm_subscriber_system.cc#L134)):*
<details><summary><i>Previously, ggould-tri wrote…</i></summary><blockquote>

FYI This will be clearer to any future, post-5725 reader if you make it clear that "now" here is a name, ie change it to now() or whatever the actual call should be (because 5725 may be something that takes a long time to resolve and you may not be the person who fixes this code then).
</blockquote></details>

Done.

---

*[drake/systems/lcm/test/lcm_subscriber_system_test.cc, line 28 at r5](https://reviewable.io:443/reviews/robotlocomotion/drake/5712#-KgzJ7VE7JcMd4snlZ2T:-KgzOViQm7m3MwS4t0xg:bvsjnls) ([raw file](https://github.com/robotlocomotion/drake/blob/891cd64b45ff65b94a1cead2da4a81ad56b53065/drake/systems/lcm/test/lcm_subscriber_system_test.cc#L28)):*
<details><summary><i>Previously, ggould-tri wrote…</i></summary><blockquote>

It is not clear to me why this is `if` rather than `while`.  Shouldn't all of `actions` be processed?
</blockquote></details>

there should only be one event in this case. i can add a demand. 

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/5712)*
<!-- Sent from Reviewable.io -->

ggould-tri commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Reviewed 2 of 2 files at r6.
Review status: all files reviewed at latest revision, 2 unresolved discussions.

---

*[drake/systems/lcm/lcm_subscriber_system.cc, line 86 at r5](https://reviewable.io:443/reviews/robotlocomotion/drake/5712#-KgzFlwaUE_i0dU5tFVz:-KgzQXy-dE_IWG-qHNwF:b32w0x) ([raw file](https://github.com/robotlocomotion/drake/blob/891cd64b45ff65b94a1cead2da4a81ad56b53065/drake/systems/lcm/lcm_subscriber_system.cc#L86)):*
<details><summary><i>Previously, siyuanfeng-tri wrote…</i></summary><blockquote>

i am not sure.. it's such a well known issue though. 
</blockquote></details>

OK

---

*[drake/systems/lcm/test/lcm_subscriber_system_test.cc, line 28 at r5](https://reviewable.io:443/reviews/robotlocomotion/drake/5712#-KgzJ7VE7JcMd4snlZ2T:-KgzQ_YBOlBJdbCHu0SX:b-4wdodv) ([raw file](https://github.com/robotlocomotion/drake/blob/891cd64b45ff65b94a1cead2da4a81ad56b53065/drake/systems/lcm/test/lcm_subscriber_system_test.cc#L28)):*
<details><summary><i>Previously, siyuanfeng-tri wrote…</i></summary><blockquote>

there should only be one event in this case. i can add a demand. 
</blockquote></details>

OK, excellent.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/5712#-:-KgzQfbHZjKJDGUBhKjy:bnfp4nl)*
<!-- Sent from Reviewable.io -->

jwnimmer-tri commented:
Reviewed 2 of 2 files at r6.
Review status: all files reviewed at latest revision, all discussions resolved, some commit checks pending.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/5712)*
<!-- Sent from Reviewable.io -->
